llvm/lib/Target/SPIRV/SPIRVBuiltins.cpp

```diff
@@ -2868,6 +2868,61 @@ static SPIRVType *getSampledImageType(const TargetExtType *OpaqueType,
   return GR->getOrCreateOpTypeSampledImage(OpaqueImageType, MIRBuilder);
 }
 
+static SPIRVType *getInlineSpirvType(const TargetExtType *ExtensionType,
+                                     MachineIRBuilder &MIRBuilder,
+                                     SPIRVGlobalRegistry *GR) {
+  assert(ExtensionType->getNumIntParameters() == 3 &&
+         "Inline SPIR-V type builtin takes an opcode, size, and alignment "
+         "parameter");
+  auto Opcode = ExtensionType->getIntParameter(0);
+
+  return GR->getOrCreateUnknownType(
+      ExtensionType, MIRBuilder, Opcode,
+      [&ExtensionType, &GR, &MIRBuilder](llvm::MachineInstrBuilder Instr) {
```

Comment:
What is the reason for the lambda? It ends up being hard to follow because you create part of the instruction here. Part in the global registry. Could the GR function take the opcode and an array_ref of registers that are suppose to be the operands?

Reply:
The reason is that not all the arguments are registers. `spirv.IntegralConstant` and passing in a normal type are lowered to register uses, but `spirv.Literal` is lowered as an immediate integer. I wasn't sure the best way to represent this – a tagged union like Rust's enum types would work well, but that doesn't seem to be as common a pattern in LLVM.